### ISSUE TITLE
Simplify the Points Hooks API

### ISSUE THREAD

**Issue by JDGrimes** [WordPoints/wordpoints#86]
The points hooks API is unnecessarily complex. We should see if we can simplify it. The current state is like this:

- Each type of points hook is represented by a class, which is an extension of `WordPoints_Points_Hook`.
- Each class is registered with `WordPoints_Points_Hooks::register()`.
- `WordPoints_Points_Hooks` calls `init()` on each registered points hook class.
- The `WordPoints_Points_Hook:init()` method calls `WordPoints_Points_Hooks::_register()` for each saved instance of the hook. If there are no saved instances of the hook, the hook is registered anyway with the ID number of `0`.
- The registered points hook instances are all stored in `WordPoints_Points_Hooks::$handlers`, as instances of the `WordPoints_Points_Hook` class indexed by hook ID.
- Which hook instances are hooked to which points types is stored in the `wordpoints_points_types_hooks` option, which is accessed through `WordPoints_Points_Hooks::get_points_types_hooks()`.

This is actually the simplified version, as I've left out the network-wide hook part. This state of affairs can actually be blamed on whoever designed the WordPress Widgets API (which is actually far more complex, would you believe). I will take the blame for not simplifying it enough, however.

There are several things I'd like to propose:

1. In the spirit of #85, we should change the way that points hook instances are registered. We should stop registering the number `0` placeholder instance, and instead store a list of all hook handlers (read: points hook type class instances) separate from the list of points hook instances.
2. We should try to simplify the way that the points type is determined. We are trying to satisfy two different needs: (1) we need to be able to get a list of all instances hooked to a points type so that we can display them to the user on the hooks screen, and (2) we need to be able to get a list of all of the instances of a type of points hook so that points can be awarded for each when it is fired.

Number 1 is an easier change that is completely internal, and doesn't involve the database.

Number 2 is more difficult, and could require database changes to really fix it. It is something that I've gone over many times before, and I never came to a solution before my mind got warped. Maybe this time will be different. :)

Also, this should be accompanied by improved documentation. Originally there should have been more differentiation between points hook types and points hook instances. It might actually have been better to name the `WordPoints_Points_Hook` class `WordPoints_Points_Hook_Type`. Even if we don't implement the `WordPoints_Points_Hook_Type` (which we actually could), we should improve this distinction in the documentation.

**Comment by JDGrimes** [comment]
> One solution for this would be to store the hook instances in their own database table, indexed by points type and hook type.

I've just thought of a problem with this idea. What to do with network hooks? We'd probably have just one table for an entire network, with columns for network ID and site ID. But for network hooks, we'd have to designate them by using the site ID 0, or in some other way.

Another thing to consider is what we'd then do with the hook settings. Would we still store them serialized in a column of the table, or would we implement an entire metadata API for them?

So yeah, I'm still thinking this might be overkill.

**Comment by JDGrimes** [comment]
> Actually, it would be pretty trivial to add an _order key to each hook instance's options.

However, we'd have to update every hook to update the order.

**Comment by JDGrimes** [comment]
> However, we'd have to update every hook to update the order.

Unfortunately, that's true for the database `order` column as well.

I found an [answer to a similar question on StackOverflow](http://stackoverflow.com/a/5504275/1924128) that suggested using a doubly linked list. The only drawback to this approach is that you can't easily, e.g., pull the first x items in the list from the database. But we'll always be using the whole list, so this might be the best way to go.

**Comment by JDGrimes** [comment]
Of course, a doubly linked list doesn't have to tie us down to a table or options for storing the hook data, we could use either. But we don't have to use a doubly linked list, we could adopt an approach similar to what we currently do with ranks. It's not like it has to scale way out, we're not going to have thousands of hooks in a single list.

As far as table vs. options, I thinking now, in light of #321, that the hooks API is maturing enough that a table might not be overkill. The main questions we need to consider are:

1. Do we need a full-fledged metadata API for hooks?
1. Should we have one network-wide table or separate tables for each site?

These two questions may actually be linked together. Since the ranks API may actually be superseded  by the new hooks API, we may want to take our cue from it. For it both of these questions are "Yes" (OK, the second one isn't really yes/no, but "network-wide"). But before we just adopt that pattern let's do a bit more research on it.

**Comment by JDGrimes** [comment]
> These two questions may actually be linked together.

What I mean by this is a single metadata table will be huge. But probably way smaller than `usermeta`, so not a deal-breaker, I guess.

The hooks don't otherwise *have* to have a meatadata API. And we could deprecate the ranks API if we need too. I think we should look at how the ranks metadata API is actually being used internally, and then consider what possible future uses we might have for such an API for hooks. But really, whether it is implemented as a table or a single column isn't necessarily a big deal for the code using the API either way, except for performing searches. I'm not sure we really need that feature though.

**Comment by JDGrimes** [comment]
We don't currently do any queries on rank meta.

**Comment by JDGrimes** [comment]
With #321 a new hooks API has now been merged into core. It does not directly address the question of how to store hooks, allowing them to be stored in any manner that a developer chooses. There is a common interface for accessing the hooks and their metadata, but each hook store can have its own implementation.

Because of this, the question of sort order was also not addressed at all, although of course it is possible to add that to the existing API without modifying any of the existing code structure (depending on how we want to handle the ordering).

As far as the concept of the complexity of the API goes, we have also addressed that. Looking back, the problem with the API wasn't so much that it was so complex, but rather confused. In the new API we have split the term hook up into different parts, and no longer actually represent it with a discrete class in the code. It is now just a general concept, the working out of which is composed of multiple parts. The "hook instances" have become "reactions", the "hook handlers" have been split into "reactors" which respond to hook fires, "reaction stores", and "events" which represent a collection of actions. These latter are the most direct descendants of the old "hook handler"/"hook type" concept.
